python/vowpalwabbit/pyvw.py

```diff
@@ -501,6 +501,25 @@ def get_weight(self, index, offset=0):
         """
         return pylibvw.vw.get_weight(self, index, offset)
 
+    def get_weight_from_name(self, feature_name, namespace_name=" "):
+        """Get the weight based on the feature name and the namespace name.
+
+        Parameters
+        ----------
+        feature_name : str
+            The name of the feature
+        namespace_name : str, by default is ""
+            The name of the namespace where the feature lives
+
+        Returns
+        -------
+        weight : float
+            Weight for the given feature and namespace name
+        """
+        space_hash = self.hash_space(namespace_name)
```

Comment:
This method should take an optional "string value" input for JSON-based inputs, and will likely need special-casing for [chain_hash](https://github.com/VowpalWabbit/vowpal_wabbit/blob/21030f5c97b71f91224bfb44c4381316cf82efeb/vowpalwabbit/parse_args.cc#L456-L460) mode.

The key bit is that in the case of string-values features in JSON, there are two behaviours: One which concatenates the value to the string, and one which uses the hash-chaining trick. This currently defaults to former and lets you switch to the latter, but this is intended to change in a future version.

Reply:
The wiki also mentions this: 

https://github.com/VowpalWabbit/vowpal_wabbit/wiki/Input-format

> Features is a sequence of whitespace separated strings, each of which is optionally followed by a float (e.g., NumberOfLegs:4.0 HasStripes) or a string (e.g., city:paris). Each string is a feature and the value is the feature value for that example. Omitting a feature means that its value is zero. Including a feature but omitting its value means that its value is 1. When a string is supplied as the feature value, the value is 1 and the feature index is calculated as hash(feature_value, hash(feature_name, namespace_hash)) where hash's signature is hash(input, seed). This chained hashing is denoted by a ^ in the audit output.

Reply:
> This method should take an optional "string value" input for JSON-based inputs, and will likely need special-casing for chain_hash mode.
>
> The key bit is that in the case of string-values features in JSON, there are two behaviours: One which concatenates the value to the string, and one which uses the hash-chaining trick. This currently defaults to former and lets you switch to the latter, but this is intended to change in a future version.

I am not sure how to integrate it since I have never used JSON-based inputs. Could you provide me with an example using the method `vw.get_weight`. Thanks !

Reply:
Chain hash actually works in text format too. It is for the use case of text feature values when you still want to use `:`. Activated with `--chain_hash` it would look like:
```
|test name:value
```

Would result in a feature index of:
```c++
hash("value", hash("name", hash("test", hash_seed)))

// Where hash("test", hash_seed) is the same as self.hash_space("test")
```

I think it could be reasonably modeled as the function taking an optional `feature_value_string` parameter and doing the chain hash in that case and if that parameter is `None` then using your current impl

Reply:
The JSON analogue of your VW text example `1 | a a b c |ns x` is:


```json
{
   "_label": 1,
   "a": 2,
   "b": 1,
   "c": 1,
   "ns": 
   {
      "x": 1
   }
}
```

Not sure if is legal to have two of the same key in a single json object, so I combined them - thus they are not 100% equivalent.

You would also need to pass the "--json" flag to the vw object on init to switch the parser into JSON mode.

Reply:
I try to parse this example using `--json` flag: 

```python
ex_json = {"_label": 1, "a": 2, "b": 1, "c": 1, "ns": {"x": 1}}
ex_json_str = str(ex_json)

model = vw("--json")
model.parse(ex_json_str)
```
but I get the following error : `RuntimeError: JSON parser error at 1: Missing a name for object member.. Handler: State: Default`

Reply:
Let me take a look at what is going on here with the parsing, but it seems like you should be able to test chain-hash mode in text mode as well, as per @jackgerrits in the meanwhile

Reply:
@etiennekintzler if you try `model.parse(json.dumps(ex_json))` should work

Also we can merge this now and open a ticket for chain hash or do you prefer to have it go in with this PR?

Reply:
You are right @olgavrou `json.dumps(ex_json)` does work. The difference with `ex_json_str` is the presence of double quote for keys (which is JSON valid) vs single quote for the cast `str(ex_json)`.

Regarding chain hash, I don't feel competent tackling it since I haven't used the functionnality.  Based on what @jackgerrits said I tried the following code but didn't work out (weight is 0.0):

```python
from vowpalwabbit.pyvw import vw

model = vw("--chain_hash")
for ex in ["1 |test name:value"]:
    model.learn(ex)

ns_hash = model.hash_space("test")
key_hash = model.hash_feature("name", ns_hash)
value_hash = model.hash_feature("value", key_hash)
model.get_weight(value_hash) # 0.0
```

I am using the last vw version (8.11.0). Do you also have the same result ?

Otherwise, I think you can merge this PR if it looks good to you !

Reply:
Looks right to me at first glance so I'm not sure what's up there. I think we should merge as this is super useful! Then someone can tackle chain hash separately with an optional argument

Reply:
created issue to track this #3081

Reply:
also yes @etiennekintzler tried it on master and I get the same as you

Reply:
chain hash applied the parse mask which is probably why the above isn't working see [here](https://github.com/VowpalWabbit/vowpal_wabbit/blob/master/vowpalwabbit/vw.h#L203)

Reply:
Looks like the plan is to deal with chain-hash after getting this PR in, right @olgavrou, @etiennekintzler? If so, looks good.

Reply:
@lokitoth yup see #3081